### PR TITLE
Gate.io Websocket Implementation

### PR DESCRIPTION
**Pull request by gmax0**

### CODE REVIEW

File: xchange-stream-gateio/src/main/java/GateioStreamingService.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+   * @param args CurrencyPair to subscribe to, followed by the refresh interval (Integer), followed
+   *     by the depth (Integer, only applicable for OrderBook channel subscriptions)
+   * @return
+   * @throws IOException
+   */
+  @Override
+  public String getSubscribeMessage(String channelName, Object... args) throws IOException {
+    final CurrencyPair currencyPair =
+        (args.length > 0 && args[0] instanceof CurrencyPair) ? ((CurrencyPair) args[0]) : null;
+    final Object[] argz =
```
Comment (earce) [inline xchange-stream-gateio/src/main/java/GateioStreamingService.java:116]: is there not a more elegant way to do this then nested args arr?

File: xchange-stream-gateio/src/main/java/GateioStreamingService.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+            : null;
+    final Integer depth =
+        (argz != null && argz.length > 1 && argz[1] instanceof Integer)
+            ? ((Integer) argz[1])
+            : null;
+
+    String baseChannelName = channelName.split(CHANNEL_NAME_DELIMITER)[0];
+
+    GateioWebSocketSubscriptionMessage subscribeMessage =
+        new GateioWebSocketSubscriptionMessage(baseChannelName, currencyPair, interval, depth);
```
Comment (earce) [inline xchange-stream-gateio/src/main/java/GateioStreamingService.java:130]: inline this with line below

File: xchange-stream-gateio/src/main/java/GateioStreamingService.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+  }
+
+  @Override
+  protected WebSocketClientExtensionHandler getWebSocketClientExtensionHandler() {
+    return WebSocketClientCompressionAllowClientNoContextAndServerNoContextHandler.INSTANCE;
+  }
+
+  @Override
+  public String getUnsubscribeMessage(String channelName) throws IOException {
+    return null;
```
Comment (earce) [inline xchange-stream-gateio/src/main/java/GateioStreamingService.java:142]: does this not exist?

File: xchange-stream-gateio/src/main/java/dto/response/GateioTradesResponse.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    return new CurrencyPair(result.getCurrencyPair().replace('_', '/'));
+  }
+
+  public Trade toTrade() {
+    return new Trade(
+        Order.OrderType.ASK,
+        new BigDecimal(result.amount),
+        getCurrencyPair(),
+        new BigDecimal(result.price),
+        Date.from(Instant.ofEpochMilli(new BigDecimal(result.createTimeMilliseconds).longValue())),
```
Comment (earce) [inline xchange-stream-gateio/src/main/java/dto/response/GateioTradesResponse.java:57]: can you not just do new Date(new BigDecimal(result.createTimeMilliseconds).longValue())

File: xchange-stream-gateio/src/main/java/dto/response/GateioOrderBookResponse.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                        currencyPair,
+                        Long.toString(result.lastUpdateId),
+                        null,
+                        new BigDecimal(rawBid[0])))
+            .collect(Collectors.toList());
+
+    List<LimitOrder> asks =
+        Arrays.stream(result.asks)
+            .map(
+                rawBid ->
```
Comment (earce) [inline xchange-stream-gateio/src/main/java/dto/response/GateioOrderBookResponse.java:67]: you should use rawAsk for better naming

File: xchange-stream-gateio/src/main/java/dto/response/GateioOrderBookResponse.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                    new LimitOrder(
+                        Order.OrderType.BID,
```
Comment (earce) [inline xchange-stream-gateio/src/main/java/dto/response/GateioOrderBookResponse.java:69]: ask?

File: xchange-stream-gateio/src/main/java/dto/response/GateioOrderBookResponse.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+                        new BigDecimal(rawBid[1]),
+                        currencyPair,
+                        Long.toString(result.lastUpdateId),
+                        null,
+                        new BigDecimal(rawBid[0])))
+            .collect(Collectors.toList());
```
Comment (earce) [inline xchange-stream-gateio/src/main/java/dto/response/GateioOrderBookResponse.java:75]: for these should your tests assert bids vs asks

File: xchange-stream-gateio/src/main/java/GateioStreamingService.java (hunk 1 above)
Comment (9-9-0) [reply xchange-stream-gateio/src/main/java/GateioStreamingService.java:116]: I can't seem to sneak spaghetti code through you, I'll see if I can clean this up a bit.

Comment (earce) [reply xchange-stream-gateio/src/main/java/GateioStreamingService.java:116]: @9-9-0 don't you prefer a good pair of eyes ;)

File: xchange-stream-gateio/src/main/java/dto/response/GateioTradesResponse.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  @Override
+  public CurrencyPair getCurrencyPair() {
+    return new CurrencyPair(result.getCurrencyPair().replace('_', '/'));
+  }
+
+  public Trade toTrade() {
+    return new Trade(
+        Order.OrderType.ASK,
```
[the remark this answers is not in the document]
Comment (gmax0) [reply xchange-stream-gateio/src/main/java/dto/response/GateioTradesResponse.java:53]: Misread the documentation, trades events are always notified on taker execution matches. Read that as trades are always asks somehow...brb retaking English courses.

**Comment by gmax0** [comment]
@earce reopening PR, rebased something incorrectly.

**Pull request closed without merging**